src/libclipper/include/clipper/future.hpp

```diff
@@ -0,0 +1,105 @@
+#ifndef CLIPPER_LIB_FUTURE_HPP
+#define CLIPPER_LIB_FUTURE_HPP
+
+#include <atomic>
+#include <cassert>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include "boost/thread.hpp"
+
+namespace clipper {
+
+namespace future {
+
+/*
+ * The first element in the pair is a future that will complete when
+ * all the futures provided have completed. The second element is a vector
+ * of futures that have the same values and will complete at the same time
+ * as the futures passed in as argument.
+ *
+ * \param futures A list of futures to compose
+ * \param num_completed A counter indicating the number of futures that
+ * have completed so far. The reason this is an argument to the function
+ * instead of internal state is that the caller of `when_all` must make sure
+ * that the counter stays in scope for the lifetime of all the futures.
+*/
+template <class T>
+std::pair<boost::future<void>, std::vector<boost::future<T>>> when_all(
+    std::vector<boost::shared_future<T>> futures,
+    std::shared_ptr<std::atomic<int>> num_completed) {
+  if (futures.size() == 0) {
+    return std::make_pair(boost::make_ready_future(),
+                          std::vector<boost::future<T>>{});
+  }
+  int num_futures = futures.size();
+  auto completion_promise = std::make_shared<boost::promise<void>>();
+  std::vector<boost::future<T>> wrapped_futures;
+  for (auto f = futures.begin(); f != futures.end(); ++f) {
+    wrapped_futures.push_back(f->then(
```

Comment:
Based on my understanding, it seems that invoking the continuation function `then` results in the creation of an additional thread when the future `f` is complete. This thread executes the supplied method and is quickly destroyed.

If we were to execute `when_all` on a large number of futures that all complete around the same period of time, wouldn't we be prone to a large spike in thread creations that may still pose a problem in the face of a ulimit? The alternative, inefficient boost implementation seems to guarantee that only a single additional thread is created per batch of futures, whereas we'd create X threads for a call to the new `when_all()` implementation with X futures supplied, albeit at different times. 

Please let me know if there's something I'm missing / provide your assessment of this tradeoff if it's accurate.